src/signer.rs

```diff
@@ -163,6 +199,43 @@ impl Signer {
     }
 }
 
+impl Secret {
+    fn sign(&self, _signing_input: &String) -> Result<Vec<u8>, SignerError> {
+        match self {
+            #[cfg(feature = "openssl")]
+            Secret::OpenSSL(key) => {
+                let signing_input = _signing_input;
+                let mut signer = SslSigner::new(MessageDigest::sha256(), key)?;
+                signer.update(signing_input.as_bytes())?;
+                let signature_payload = signer.sign_to_vec()?;
+                Ok(signature_payload)
+            }
+            #[cfg(all(not(feature = "openssl"), feature = "ring"))]
+            Secret::Ring(key) => {
+                #[cfg(feature = "ring")]
+                let signing_input = _signing_input;
+                let rng = rand::SystemRandom::new();
+                let signature_payload = key.sign(&rng, signing_input.as_bytes())?;
+                Ok(signature_payload.as_ref().to_vec())
```

Comment:
`ring` has a trait `SecureRandom`, which is implemented by https://docs.rs/ring/0.16.20/ring/rand/struct.SystemRandom.html. They use a secure system-provided random generator.

However, now reading the docs I see two other problems:

1. I used a deprecated method `sign`,
2. Docs of `ring` suggest not to create `SecureRandom` all the time, but instead pass it as param.

Reply:
Good idea. The random generator is now stored in the `Secret`. I asked about the second point on `ring`'s issue tracker. It might take a while until they answer. I don't see how not to use this method atm.